### PR TITLE
Focus Shift to Pipelines Best Practices

### PR DESCRIPTION
**Pull request by kevinlibuit**
This PR aims to shift the guidance document focus from proposed standards for bioinformatics software to best practices for bioinformatics pipelines. This shift in focus helps to highlight recommendations for the specific type of software that public health labs are having challenges with implementing for pathogen genomics, i.e. bioinformatics workflows; the shift from proposed standards to best practices aims to reflect the evolving nature of these technologies and encourage continuous flexibility, consensus building, and more rapid adoption throughout the community.

The changes to this document include:
1. Shift in language from `Proposed Standards` to `Best Practices`
2. Narrowed scope from `bioinformatics software` to `bioinformatics pipelines`
   a. Language was added in the Overview section to define this subcategory of bioinformatics software
3. Removal of `Software Security and Vulnerabilities` section (refer to #33 for rationale) 
4. Adding of `Workflow Management System` -- a best practice specific to bioinformatics pipelines
4. Reorganization into 10 Best Practices with "Pipeline Documentation" encompassing several documentation practices:
``` 
1. Publicly-Accessible Repository
2. Open-Source License
3. Version Controlled Software
4. Workflow Management System
5. Containerized/Packaged Software
6. Common File Formats
7. Software Testing
8. Benchmark/Validation Datasets
9. Reference Data Requirements
10. Pipeline Documentation
    10a. Statement of Need with Respect to Public Health Pathogen Genomics
    10b. Pipeline Functionality
    10c. Contribution, Authorship, and Verified Point of Contact
    10d. Conflict of Interest Statement
    10e. Community Guidelines for Contribution and Support
    10f. Documentation for Local Installation and/or Remote Access (e.g. Web Server or Galaxy/Terra Workflow)
    10g. Example Usage

### CODE REVIEW

**Comment by kevinlibuit** [comment]
Updated the `Pipeline Documentation` practice to include `Pipeline Maintenance Statement`. I've also reordered this section as follows: 
```
10. Pipeline Documentation
    10a. Contribution, Authorship, and Verified Point of Contact
    10b. Conflict of Interest Statement
    10c. Pipeline Maintenance
    10d. Community Guidelines for Contribution and Support
    10e. Statement of Need with Respect to Public Health Pathogen Genomics
    10f. Pipeline Functionality
    10g. Documentation for Local Installation and/or Remote Access (e.g. Web Server or Galaxy/Terra Workflow)
    10h. Example Usage
    
```

File: docs/pipeline-standards.md [hunk 1]
```diff
@@ -15,241 +15,244 @@ Libuit KG, Guthrie J, Ambrosio F, Kapsak C, Unal Gultekin, Holmes J, Wright S, N
```
Comment (svarona) [inline docs/pipeline-standards.md:15]: Maybe we should add these changes to the changelog?

File: docs/pipeline-standards.md [hunk 2]
```diff
@@ -15,241 +15,244 @@ Libuit KG, Guthrie J, Ambrosio F, Kapsak C, Unal Gultekin, Holmes J, Wright S, N
...
 
-## Proposed Standards
+## 10 Best Practices for Public Health Bioinformatics Pipelines
 
-### Publicly-Accessible Repository
-_Is the source code for this software available at a publicly-accessible repository URL?_
+### 1. Publicly-Accessible Repository
+_Is the source code for this pipeline available at a publicly-accessible repository URL?_
 
 Publicly-accessible software bolsters collaboration and expedites innovation in public health bioinformatics, empowering worldwide public health communities to address critical challenges. By enhancing accessibility, publicly available software enable interoperability and reproducibility across public health investigations, crucial for well-informed decision-making and policy creation. Popular code repositories such as GitHub, GitLab, Bitbucket, and SourceForge offer platforms for developers to share their work.
```
Comment (svarona) [inline docs/pipeline-standards.md:29]: should we change software to pipelines or workflows here?

File: docs/pipeline-standards.md [hunk 3]
```diff
@@ -15,241 +15,244 @@ Libuit KG, Guthrie J, Ambrosio F, Kapsak C, Unal Gultekin, Holmes J, Wright S, N
...
+Accepting and generating common file formats for public health bioinformatics pipeline enhances interoperability and data exchange between different tools and platforms. This approach facilitates data sharing, promotes consistency, and enables researchers to leverage a wide range of pipeline solutions, contributing to the development of more comprehensive and effective solutions for addressing critical public health challenges. These files include: .fasta, .fastq(.gz), .sam, .bam, .bai, .bed, .vcf, .gff, .gtf, .txt, .log, .tsv, .csv, .nwk, and .json.
```
Comment (svarona) [inline docs/pipeline-standards.md:87]: These files include among others (i miss some of them like .hmm, .tbi for vcf files). Should we generate a file with a complete list of formats that would be updated in long term?

File: docs/pipeline-standards.md [hunk 4]
```diff
@@ -15,241 +15,244 @@ Libuit KG, Guthrie J, Ambrosio F, Kapsak C, Unal Gultekin, Holmes J, Wright S, N
...
+**To adhear this best practice:** Implement pipeline components within Docker containers or distribute them as Conda packages.
```
Comment (svarona) [inline docs/pipeline-standards.md:80]: I will add here something related to the `There should be a clear summary in the Git README`

File: docs/pipeline-standards.md [hunk 5]
```diff
@@ -15,241 +15,244 @@ Libuit KG, Guthrie J, Ambrosio F, Kapsak C, Unal Gultekin, Holmes J, Wright S, N
...
 
-- To meet this standard, authors of public health bioinformatics software, pipelines, or methods should transparently disclose any potential conflicts of interest, such as commercial, personal, political, or religious affiliations, to promote transparency and integrity in scientific research.
-- To verify this standard, the reviewer should assess the presence of a clear conflict of interest statement in the documentation or repository, ensuring that responsible authors have openly disclosed any factors that may impact their ability to perform an unbiased peer review or evaluation.
+**To verify adherence to this best practice:** The reviewer should inspect the pipeline source code and review analyatical steps (e.g. NextFLow processes or WDL tasks) to ensure use of containerized or packaged softare.
```
Comment (svarona) [inline docs/pipeline-standards.md:82]: Also, if we say before that `there should be a supplementary document justifying, where appropriate, why a particular version was chosen for analysis`, maybe we need to specify here that the reviewer should ensure that document exist?

File: docs/pipeline-standards.md [hunk 6]
```diff
@@ -15,241 +15,244 @@ Libuit KG, Guthrie J, Ambrosio F, Kapsak C, Unal Gultekin, Holmes J, Wright S, N
...
+There should be a clear summary in the Git README pointing to which containerisation method has been chosen and instructions for how a lab can install this pipeline and/or where docker images are available, e.g. dockerhub and quay for containers; and where conda packages are available, e.g. anaconda and bioconda (cross-referenced with Installation Instructions).
 
-### Conflict of Interest Statement
-_Have all potential conflicts of interest been disclosed?_
+When bioinformatics software is containerised, there should be a supplementary document justifying, where appropriate, why a particular version was chosen for analysis. Reasons for choosing specific pipeline versions include but are not limited to ensuring functionality by selecting compatible dependencies and major differences in functionality and features between pipeline versions.
```
Comment (svarona) [inline docs/pipeline-standards.md:77]: Maybe it is not necessary to justify a specific version if it was the latest version when the pipeline release was done?

File: docs/pipeline-standards.md [hunk 7]
```diff
@@ -15,241 +15,244 @@ Libuit KG, Guthrie J, Ambrosio F, Kapsak C, Unal Gultekin, Holmes J, Wright S, N
...
+- Ensure that the database is compatible with the pipeline that it is being used with. Make sure that the documentation clearly states the versions of pipeline and systems that the database is compatible with.
 
-```
-Input files:
-- sample.vcf (Variant Call Format file containing genomic variants)
+The format of open-access downloadable should be defined, ideally in compressed format, and in such a format that will be best suited for downstream usage/analysis.
 
-Command:
-$ analyze_variants.py -i sample.vcf -o output.txt -p population_data.csv
+**To adhere to this best practice:** Tthe required reference data and/or databases for public health bioinformatics pipeline should be clearly documented, publicly accessible, and maintained. If an external reference database is required, it should also adhere to standards such as static versioning, open-access, and clear instructions for installation/access. Clearly document aspects of the database, such as the schema, table structure, and stored procedures, in a format suitable for users, such as technical documentation, user guides, and reference manuals.
```
Comment (svarona) [inline docs/pipeline-standards.md:142]: Tthe typo

File: docs/pipeline-standards.md (hunk 6 above)
Comment (svarona) [reply docs/pipeline-standards.md:77]: I think it's not clear here if you are speaking about software version or pipeline version. I'm a little bit confused.

File: docs/pipeline-standards.md [hunk 8]
```diff
@@ -15,241 +15,244 @@ Libuit KG, Guthrie J, Ambrosio F, Kapsak C, Unal Gultekin, Holmes J, Wright S, N
...
 
-Including software tests for public health bioinformatics software ensures the reliability and accuracy of the tools, enhancing user confidence and promoting consistent research outcomes. These tests also facilitate early detection and resolution of potential issues, contributing to the overall stability and robustness of the software in a rapidly evolving public health landscape. At a minimum, software being implemented for public health pathogen genomics should include: 
-- Smoke tests to ensure that the basic functionality of the program is working correctly
-- Unit tests to test individual code units
-- System tests/end-to-end tests to assess the overall functionality of the program, with a focus on common and important paths
-- Regression tests to ensure that changes to the code do not break existing functionality (note: system tests can be utilized to implement regression tests)
+**To adhere to this best practice and verify adherence to this best practice:** Refer to the documentation practices listed below.
```
Comment (svarona) [inline docs/pipeline-standards.md:158]: Maybe we should explain how this Pipeline Documentation is going to be evaluated. Is it going to be mandatory to pass `10. Pipeline Documentation` to have all 10a to 10h passed before?

File: docs/pipeline-standards.md [hunk 9]
```diff
@@ -15,241 +15,244 @@ Libuit KG, Guthrie J, Ambrosio F, Kapsak C, Unal Gultekin, Holmes J, Wright S, N
...
 
+**To adhere to this documentation practice:** Clearly list authorship and credit for the bioinformatics tool, acknowledging individual contributions and following the CRedIT system for appropriate recognition based on specific contributions.
 
-### Community Guidelines for Contribution and Support
-_Are there clear guidelines for third parties wishing to 1) contribute to the software 2) report issues or problems with the software and 3) seek support?_
+**To verify adherence to this documentation practice:** The reviewer should check the documentation and repository for a clear and comprehensive list of authors, ensuring adherence to the CRedIT system and appropriate acknowledgment of contributors based on their roles in programming, pipeline development, design, implementation, and testing.
```
Comment (svarona) [inline docs/pipeline-standards.md:169]: If to verify this we are going to say `ensuring adherence to the CRedIT system`, maybe in the text we should change `We would recommend using the [CRedIT system]`. It is not a recommendation if you require it.

File: docs/pipeline-standards.md [hunk 10]
```diff
@@ -15,241 +15,244 @@ Libuit KG, Guthrie J, Ambrosio F, Kapsak C, Unal Gultekin, Holmes J, Wright S, N
...
 
-- To meet this standard, include a benchmark or validation dataset for public health bioinformatics software, promoting transparency and consistency in the evaluation and comparison of different tools. 
-- To verify this standard, the reviewer should check the documentation or repository for information on the availability and accessibility of a benchmark and/or validation dataset for the public health bioinformatics software.
+This documentation may detail the support mechanism for users, including how they can report issues, request features, or contribute to the project. If applicable, this documentation may also cover the funding model or community support strategy that will ensure the pipeline's ongoing development and maintenance. This could include details on how the project is funded, plans for seeking future funding, or how the project fosters a community of contributors.
```
Comment (svarona) [inline docs/pipeline-standards.md:185]: 💯

File: docs/pipeline-standards.md [hunk 11]
```diff
@@ -15,241 +15,244 @@ Libuit KG, Guthrie J, Ambrosio F, Kapsak C, Unal Gultekin, Holmes J, Wright S, N
...
 
-### Container/Packaged Software
-_Has the software been made available through a packaging manager, e.g. conda, or software container, e.g. Docker, to enhance interoperable software distribution?_ 
+**To adhere to this documentation practice:** Authors must include a detailed statement regarding the pipeline's future upkeep, covering aspects such as update schedules, user support mechanisms, and funding or community support strategies. 
 
-Using software packages and/or containerization for public health bioinformatics software enhances interoperability by enabling seamless integration and deployment across different platforms and environments. This approach simplifies software distribution and installation, promotes reproducibility, and facilitates collaboration among researchers, contributing to the development of more accessible and interoperable tools and resources.
+*To verify adherence to this documentation practice:** the reviewer will check that the documentation contains a comprehensive statement detailing the authors' commitment to maintaining the pipeline, including specific plans for updates, user support, and securing the pipeline's sustainability.
```
Comment (svarona) [inline docs/pipeline-standards.md:189]: one * missing at the beginning of the line

File: docs/pipeline-standards.md [hunk 12]
```diff
@@ -15,241 +15,244 @@ Libuit KG, Guthrie J, Ambrosio F, Kapsak C, Unal Gultekin, Holmes J, Wright S, N
...
 
-### Reference Data Requirements
-_Are required reference data and/or databases clearly documented, publicly accessible, and maintained?_
+**To adhere to this documentation practice:** Clearly state the significance and relevance of public health bioinformatics pipeline and explain how the tool differs from existing ones and/or improves established methods.
 
-Documenting any external reference data or database requirements for public health bioinformatics software enhances the usability and reproducibility of the software by providing clear and comprehensive information on the necessary data sources and dependencies. This documentation promotes transparency, facilitates replication, and enables researchers to more effectively integrate the software into their workflows, contributing to the development of more reliable and impactful tools and resources.
+**To verify adherence to this documentation practice:**the reviewer will ensure that the documentation includes a statement highlighting the pipeline's purpose and its alignment with pressing public health challenges.
```
Comment (svarona) [inline docs/pipeline-standards.md:210]: missing space between * and `the`

File: docs/pipeline-standards.md (hunk 6 above)
Comment (kevinlibuit) [reply docs/pipeline-standards.md:77]: Ah, thanks for catching this. Updating the language to clarify :

`There should be a clear summary in the Git README pointing to which containerisation method has been chosen and instructions for how a lab can install this pipeline and/or where docker images are available, e.g. dockerhub and quay for containers; and where conda packages are available, e.g. anaconda and bioconda (cross-referenced with Installation Instructions). Documentation should indicate the specific version included in the pipeline. This is important as specific software versions  may impact functionality.`

File: docs/pipeline-standards.md (hunk 8 above)
Comment (kevinlibuit) [reply docs/pipeline-standards.md]: Added clarification

**Pull request merged**